influxdb3_write/src/write_buffer/mod.rs

```diff
@@ -62,67 +68,110 @@ pub struct WriteRequest<'a> {
 #[derive(Debug)]
 pub struct WriteBufferImpl<W> {
     catalog: Arc<Catalog>,
-    buffered_data: RwLock<HashMap<String, DatabaseBuffer>>,
+    segment_state: Arc<RwLock<SegmentState>>,
     #[allow(dead_code)]
     wal: Option<Arc<W>>,
+    write_buffer_flusher: WriteBufferFlusher,
 }
 
-impl<W: Wal> WriteBufferImpl<W> {
-    pub fn new(catalog: Arc<Catalog>, wal: Option<Arc<W>>) -> Self {
+#[derive(Debug)]
+struct SegmentState {
+    open_segment: OpenBufferSegment,
+    #[allow(dead_code)]
+    persisting_segments: Vec<ClosedBufferSegment>,
+}
+
+impl SegmentState {
+    pub fn new(open_segment: OpenBufferSegment) -> Self {
         Self {
+            open_segment,
+            persisting_segments: vec![],
+        }
+    }
+}
+
+impl<W: Wal> WriteBufferImpl<W> {
+    pub fn new(
+        catalog: Arc<Catalog>,
+        wal: Option<Arc<W>>,
+        next_segment_id: SegmentId,
+    ) -> Result<Self> {
+        let segment_writer = match wal
+            .as_ref()
+            .map(|w| w.open_segment_writer(next_segment_id))
+            .transpose()?
+        {
+            Some(writer) => writer,
+            None => Box::new(WalSegmentWriterNoopImpl::new(next_segment_id)),
+        };
```

Comment:
You can replace this with the [`unwrap_or_else`](https://doc.rust-lang.org/std/option/enum.Option.html#method.unwrap_or_else) function and it gets rid of the need for a `match`.

```suggestion
            .unwrap_or_else(|| Box::new(WalSegmentWriterNoopImpl::new(next_segment_id)));
```